src/Stratis.Bitcoin/P2P/PeerAddressManager.cs

```diff
@@ -158,7 +158,13 @@ public void SavePeers()
                 return;
 
             var fileStorage = new FileStorage<List<PeerAddress>>(this.PeerFilePath.AddressManagerFilePath);
-            fileStorage.SaveToFile(this.peers.OrderByDescending(p => p.Value.LastConnectionSuccess).Select(p => p.Value).ToList(), PeerFileName);
+            fileStorage.SaveToFile(
+                this.peers
+                    .ToList()
+                    .OrderByDescending(p => p.Value.LastConnectionSuccess)
+                    .Select(p => p.Value)
+                    .ToList(), 
```

Comment:
You probably don't need to enumerate twice so remove the second `ToList()`

Reply:
Perhaps also for readability take the snapshot first in a new param?

Reply:
Actually looking at the implementation, `ToArray` is definitely safer. `GetEnumerator` doesn't guarantee correct snapshot and you do need second `ToList` as this is the type of the parameter SaveToFile expects

Reply:
Found a much better alternative. We can use `.Values` collection which has a lock and removes the need to use `Select(v => v.Value)`

Reply:
@dangershony can we merge now?